### PR TITLE
update message processing

### PR DESCRIPTION
**Pull request by mxyng**
this change changes the way messages are processed before handing off to the llm. there are a few areas worth mentioning:

1. messages are now a first class component of the template. template rendering will only falling back to the previous iterative template if messages is unsupported by the template. however, new models _should_ implement the previous prompt/response template for compatibility with older ollama versions
2. the generate endpoint has been updated to use messages for prompt templating but the end result should be the same
3. the chat endpoint has been updated to preprocess incoming messages

    - continuous messages of the same role are joined into a single message, separated with two newlines
    - content and image data can be interleaved by sending messages with alternating fields, e.g. 
        ```
        [
            {"role": "user", "content": "Consider the following images:"},
            {"role": "user", "images": ["<base64 image data>", "<base64 image data>"]},
            {"role": "user", "content": "What is the difference between the two images?"}
        ]
        ```
    - system messages are aggregated and prepended to the _last_ user message

### CODE REVIEW

File: server/prompt.go
Comment (bmizerany) [inline server/prompt.go:14]: Does this return a "chat prompt"? This name isn't helpful when skimming. It would also help readers to name the return vars since `string` returned by `chatPrompt` with an `[]llm.ImageData` says nothing about what I'm getting back when calling this thing.

Docs please, or new names?

File: server/prompt_test.go
Comment (bmizerany) [inline server/prompt_test.go:26]: I learned a very neat trick for this from Russ:

```go
type mock struct {
       InterfaceToMock
}

func (x mock) TheFuncIWantToMock() { ... }
````

If anything calls anything other than `TheFuncIWantToMock` there will be a panic, since `InterfaceToMock` is `nil`. :)

This removes all the code below, taking away unnecessary distracting code.

File: template/template_test.go
Comment (bmizerany) [inline template/template_test.go:216]: Giving the cases some descriptive names here would be very helpful. The inputs are complex enough to warrant maintaining some words about what is being testing.

File: server/prompt.go
Comment (bmizerany) [inline server/prompt.go:15]: Typo?

Comment (bmizerany) [reply server/prompt.go:15]: Maybe "The blarb requires a system message, so include the extracted system message ..."

File: template/template.go
Comment (BruceMacD) [inline template/template.go:106]: nit: I've found that not using anonymous functions here makes these critical paths easier to test.

Ex:
```
func toJson(v any) string {
	b, err := json.Marshal(v)
	if err != nil {
		return ""
	}

	return string(b)
}

func isLastMessage(s []*api.Message, m *api.Message) bool {
	for i := len(s) - 1; i >= 0; i-- {
		if m.Role != s[i].Role {
			continue
		}

		return m == s[i]
	}

	return false
}

func Parse(s string) (*Template, error) {
	tmpl := template.New("").Option("missingkey=zero").Funcs(template.FuncMap{
		"toJson":        toJson,
		"isLastMessage": isLastMessage,
	})
```

File: server/routes.go
Comment (mxyng) [reply server/routes.go:144]: restructured this in subsequent pr which I'll bring forward. it should make this case easier to identify

File: template/template.go
Comment (mxyng) [reply template/template.go:225]: was this possible before?

File: server/prompt.go
Comment (mxyng) [reply server/prompt.go]: added docs describing the function and its purpose. the functionality mirrors the current `ChatPrompt` whose name I took but I'm open to suggestions

File: server/prompt.go
Comment (mxyng) [reply server/prompt.go]: updated the comment so it should be clearer but if that's not the case, lmk

File: server/prompt.go
Comment (bmizerany) [inline server/prompt.go:14]: Very nice! Thank you.

File: template/template.go
Comment (mxyng) [reply template/template.go]: I pulled out the funcs so it can be unit tested but kept the functions themselves anonymous since they don't really have meaning outside of the template. I also replaced `isLastMessage` with `messages.Last(role, message)` which should be clearer

File: template/template.go
Comment (BruceMacD) [reply template/template.go:236]: I do not believe it was, so feel free to call it a limitation.

File: server/routes.go
Comment (BruceMacD) [inline server/routes.go:1149]: When model isn't specified the error code should be `400` but right now the code is `500` with this logic (same in /generate).

```
HTTP/1.1 500 Internal Server Error
Content-Type: application/json; charset=utf-8
Date: Fri, 21 Jun 2024 01:51:58 GMT
Content-Length: 29
Connection: close

{
  "error": "model is required"
}
```

**Review by BruceMacD — COMMENTED** [review]
Last comment about the response code in an case, but otherwise this is looking ready to me.

File: server/images.go
Comment (BruceMacD) [inline server/images.go:84]: This pattern is a bit confusing to me, currently:
- Pass capabilities I am looking for
- For each capability I am looking for check if the model does not have that pattern

This may be out of scope for this change

File: server/routes.go
Comment (BruceMacD) [inline server/routes.go:1228]: This still needs the "model is required" `400` header case:
```
❯ curl http://127.0.0.1:11434/api/generate -d '{}' -v
*   Trying 127.0.0.1:11434...
* Connected to 127.0.0.1 (127.0.0.1) port 11434
> POST /api/generate HTTP/1.1
> Host: 127.0.0.1:11434
> User-Agent: curl/8.4.0
> Accept: */*
> Content-Length: 2
> Content-Type: application/x-www-form-urlencoded
> 
< HTTP/1.1 500 Internal Server Error
< Content-Type: application/json; charset=utf-8
< Date: Fri, 21 Jun 2024 02:14:50 GMT
< Content-Length: 29
< 
* Connection #0 to host 127.0.0.1 left intact
{"error":"model is required"}%  
```

File: server/images.go
Comment (mxyng) [reply server/images.go:84]: This seems like a reasonable pattern to me to check if a model has certain capabilities and have it return the capabilities that are missing. I'm open to suggestions

File: server/routes.go
Comment (BruceMacD) [inline server/routes.go:1234]:
```suggestion
	case errors.Is(err, os.ErrNotExist):
		c.JSON(http.StatusNotFound, gin.H{"error": fmt.Sprintf("model %q not found, try pulling it first", name)})
	case strings.Contains(err.Error(), "model is required"):
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
```

Still need something like this for the "request with no model specified" case, it will respond with 500 otherwise.

File: server/images.go
Comment (jmorganca) [inline server/images.go:65]: This felt clearer when it returned `bool`

**Review by BruceMacD — APPROVED** [review]
This looks good, excited for this one

File: server/images.go
Comment (mxyng) [reply server/images.go:65]: this should communicate what is missing which a bool return value can't do

File: server/routes.go
Comment (jmorganca) [inline server/routes.go:76]: Passing in `caps` is weird here – shouldn't the caller do that before scheduling the runner?

File: server/routes.go
Comment (jmorganca) [inline server/routes.go:86]: This is an example of where the `missing` return value is confusing - you'd expect `Has` to return whether it the model has the capabilities or not – and it could easily be mistaken for the opposite: the list that the model "has"

File: server/images.go
Comment (jmorganca) [reply server/images.go:65]: Yes but it could be error prone (see below) – could we have the caller check in a loop?

**Comment by jmorganca** [comment]
This is looking good. Love that messages are becoming a first-class part of the templating.

File: template/template.go
Comment (jmorganca) [inline template/template.go:216]: Took me a bit to understand what this function did – do we need the `role` field here?

File: template/template.go
Comment (jmorganca) [inline template/template.go:79]: The new templating is powerful, but writing templates gets quite hard – it would be great to include a `template.md` document in `docs` that outlines how to write one. I don't imagine most Ollama users will have to, but we should still try and make it as approachable as possible.

File: template/template.go
Comment (jmorganca) [reply template/template.go:216]: Or could we look at the current message's role (not sure if we need to cross-check in any cases?)

File: server/routes.go
Comment (mxyng) [reply server/routes.go:76]: the caller doesn't have access to the model so it can't check capabilities

File: template/template.go
Comment (mxyng) [reply template/template.go:216]: > could we look at the current message's role

that was the original implementation but it seemed confusing since this isn't the last message but the last message of a particular role

File: template/template.go
Comment (mxyng) [reply template/template.go:79]: that can be a follow up once this is merged

File: template/template.go
Comment (jmorganca) [reply template/template.go:216]: The part that made it hard was in the example Mistral template. 

```
[INST] {{ if and ($.Messages.Last "user" .) $.System }}{{ $.System }}{{ printf "\n\n" }}
{{- end }}
```

I'm most worried about adding a new helper function (or in this case, a method on `.Messages` as they will be forever. While it's not as pretty, could we use standard template functions?

```
{{- range $index, $message := .Messages }}
  {{- if eq $message.Role "user" }}
    {{- if and (eq $index (sub (len $.Messages) 1)) $.Tools }}
      [AVAILABLE_TOOLS] {{ toJson $.Tools }}[/AVAILABLE_TOOLS]
    {{- end }}
    [INST]
    {{- if and (eq $index (sub (len $.Messages) 1)) $.System }}
      {{ $.System }}
    {{- end }}
    {{ $message.Content }}[/INST]
  {{- else if eq $message.Role "assistant" }}
    {{- if $message.Content }}
      {{ $message.Content }}</s>
    {{- else if $message.ToolCalls }}
      [TOOL_CALLS] [
      {{- range $message.ToolCalls }}
        {{ "{" }}"name": "{{ .Function.Name }}", "arguments": {{ toJson .Function.Arguments }}{{ "}" }}
      {{- end }}]</s>
    {{- end }}
  {{- else if eq $message.Role "tool" }}
    [TOOL_RESULTS] {{ toJson $message }}[/TOOL_RESULTS]
  {{- end }}
{{- end }}
```

File: server/prompt.go
Comment (jmorganca) [inline server/prompt.go:17]: We shouldn't be passing in `runnerRef` here as it's starting to couple the scheduling stuff into this. Could we instead pass in the fields required to calculate the template:

1. Messages
2. System
3. Template
4. Tokenizer (right now this could just be a function)

etc.. I'm open to passing in the model here or even making this a function of `model`

File: server/routes.go
Comment (jmorganca) [reply server/routes.go:76]: It did before this change

Could we load the model and check the capabilities before deciding to schedule the model here? and then pass in the `Model`? It seems like a precondition the handler should check (ideally this function goes away and we do something like `schduler.Schedule(model, opts)` from the handler which would be a lot clearer

Comment (mxyng) [reply server/routes.go:76]: maybe the function is misnamed. this doesn't schedule anything until some check including capabilities. it aggregates the various model and options related lines that chat, generate, and embeddings all have to call before scheduling a runner

File: template/template.go
Comment (jmorganca) [inline template/template.go:106]: `json`? This would follow suit with built-in format funcs like `js` `html` etc

File: template/template.go
Comment (jmorganca) [inline template/template.go:114]: I think here we can just use existing built-ins (to avoid committing to helpers in the first pass) e.g.

```
{{ range $index, $item := range .Items }}
  {{ $last := eq (len (slice $.Items $index)) 1 }}
  is it last? {{ $last }}
{{ end }}
```

While not pretty, it will always work and we can "upgrade" it to a better last helper later

File: template/template.go
Comment (jmorganca) [inline template/template.go:221]: I think this makes sense, however we should make sure to document this in a follow up PR so folks know how repeat messages are handled

**Review by jmorganca — COMMENTED** [review]
Overall looks great- small comment on the templating so we don't overcommit to helpers yet unless we really have to

File: template/template.go
Comment (mxyng) [reply template/template.go]: this isn't used until tools so taking it out for now

**Review by jmorganca — APPROVED** [review]

**Comment by jmorganca** [comment]
LGTM. Will test this a bunch this evening (no need to block merging) but it looks great

**Pull request merged**
